src/Type/Php/StrCaseFunctionsReturnTypeExtension.php

```diff
@@ -0,0 +1,82 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\ShouldNotHappenException;
+use PHPStan\Type\Accessory\AccessoryLiteralStringType;
+use PHPStan\Type\Accessory\AccessoryNonEmptyStringType;
+use PHPStan\Type\Accessory\AccessoryNumericStringType;
+use PHPStan\Type\Constant\ConstantStringType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\IntersectionType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use function count;
+use function in_array;
+use function is_callable;
+
+class StrCaseFunctionsReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return in_array($functionReflection->getName(), [
+			'strtoupper',
+			'strtolower',
+			'mb_strtoupper',
+			'mb_strtolower',
+			'lcfirst',
+			'ucfirst',
+			'ucwords',
+		], true);
+	}
+
+	public function getTypeFromFunctionCall(
+		FunctionReflection $functionReflection,
+		FuncCall $functionCall,
+		Scope $scope,
+	): Type
+	{
+		$args = $functionCall->getArgs();
+		if (count($args) < 1) {
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+		$argType = $scope->getType($args[0]->value);
+
+		if ($argType instanceof ConstantStringType) {
+			$fnName = $functionReflection->getName();
+			if (!is_callable($fnName)) {
+				throw new ShouldNotHappenException();
+			}
+			return new ConstantStringType($fnName($argType->getValue()));
+		}
+
+		if ($argType->isNumericString()->yes()) {
+			return new IntersectionType([
+				new StringType(),
+				new AccessoryNumericStringType(),
+			]);
+		}
+
+		$accessoryTypes = [];
+		if ($argType->isNonEmptyString()->yes()) {
+			$accessoryTypes[] = new AccessoryNonEmptyStringType();
+		}
+		if ($argType->isLiteralString()->yes()) {
+			$accessoryTypes[] = new AccessoryLiteralStringType();
```

Comment:
I thought about the same thing and was not sure. but then I found https://github.com/phpstan/phpstan-src/blob/28f45033eb5e48bcfa6daec1f1ee053626dbc8c7/src/Type/Php/ImplodeFunctionReturnTypeExtension.php#L73-L75 which made me think this one here is fine.

maybe https://github.com/phpstan/phpstan-src/blob/28f45033eb5e48bcfa6daec1f1ee053626dbc8c7/src/Type/Php/ImplodeFunctionReturnTypeExtension.php#L73-L75 is also wrong?

Reply:
`implode` is one of the four functions that can produce literal strings as blessed by the original RFC: https://wiki.php.net/rfc/is_literal

Reply:
Yep, sorry, the only modification you can do that keep the `literal-string` flag is concatenation (this was done to help adoption, and has is seen as acceptable by several security people, like the developers behind go-safe-html; where `str_repeat()`, `str_pad()`, `implode()`, and `join()` are all glorified forms of concatenation)... other modifications are no longer result in a "fixed value in source code" ([ref](https://en.wikipedia.org/wiki/Literal_(computer_programming))), and while the security properties of some functions ([like strtoupper](https://wiki.php.net/rfc/is_literal#other_functions)) might be ok, I would like to avoid the discussion of what functions are allowed vs what aren't :-)

Reply:
thx for the insights